### ISSUE TITLE
Implement parallelism / threading capability in plans

### ISSUE THREAD

**Issue by reidmv** [puppetlabs/bolt#2127]
## Use Case

In the early days of Bolt, plan execution was entirely procedural and each procedure needed to finish before the next could start. This meant performing similar-but-not-identical actions across a large set of nodes could be very slow, as it effectively needed to be serial.

`run_task_with` provided a means of quickly running a _task_ across a set of nodes which all needed slightly different parameters. Provided you only needed to run one task with differing parameters and parallelism, this was useful. Leveraging it can result in code that [isn't very readable though](https://github.com/puppetlabs/puppetlabs-peadm/blob/fe75751b7ed35c652e13b1c9d4edf2292e75305c/plans/action/install.pp#L309-L343), compared to what it would simplistically look like to call `run_task()` multiple times with different parameters.

Suppose you want to run a sequence of tasks though, or a plan, across a set of nodes, with slightly different parameters for each. Today this requires you to go back to serial execution one node at a time, waiting for each node to finish before moving on to the next.

## Describe the Solution You Would Like

Provide real asynchronous constructs in the Bolt plan language to accelerate execution of more complicated actions across nodes in parallel.

Suggested implementation:

* `run_task_async` function
* `run_plan_async` function
* `AsyncResultSet` type
* `AsyncPlanResult` type

The `*_async` functions should immediately return their respective `AsyncResult*` type, and begin async processing of the plan or task in a separate thread.

The `AsyncResult*` types should support `.get()` methods, which when called will block until the separate thread completes and return a `ResultSet`, `PlanResult`, or `Error` object.

```puppet
# Kick off arbitrarily complex stuff you need to do in the background
$async_plan_results = $targets.each |$target| {
  run_plan_async('plan::mcplanface', $target,
    important_parameter => $hash[$target],
  }
}

# Do other stuff if you need to

# When you need the plan results
$plan_results = $async_plan_results.map |$async_result| { $async_result.get }
```

**Comment by adreyer** [comment]
We've talked about this for a long time and it seems reasonable to do some exploratory work to figure out exactly how hard it is.  We should do a spike to experiment with parallel plan execution and discover and see how we need to manage the puppet compiler.

Scope:
Implement a `plan_future` function that starts a plan and returns a `Future` datatype. The `Future` datatype should allow the plan author to:
- check the status of a future
- block until the future is resolved
- fetch the result of the future(or raise its error)
- return an error raised in the future without raising and catching it.

**Comment by reidmv** [comment]
Etymology note: the suggested `*_async` function names and general interface in the ticket description was borrowed from the [Cassandra ruby driver asynchronous IO implementation](https://docs.datastax.com/en/developer/ruby-driver/3.2/features/asynchronous_io/).

**Comment by lucywyman** [comment]
Oh my god Github you suck. Closed by #2190
